doc/man/gnunet-tbench.1
.TH GNUNET-TBENCH "1" "23 Dec 2006" "GNUnet"
.SH NAME
gnunet\-tbench \- transport profiling tool
.SH SYNOPSIS
.B gnunet\-tbench
[\fIOPTIONS\fR]
.SH DESCRIPTION
gnunet\-tbench can be used to test the performance of the GNUnet
core (link\-to\-link encryption and the available transport services).
gnunet-tbench is useless to most ordinary users since its primary
function is to test the performance and correctness
of GNUnet transport service implementations.
.PP
gnunet\-tbench sends a sequence of messages to another peer that
has the tbench module loaded. The service then measures the
throughput, latency and loss of the messages round-trip.
gnunet\-tbench can only be used to test a direct peer\-to\-peer
connection. You must load the \fItbench\fR module (via
the configuration gnunetd.conf, section GNUNETD under APPLICATIONS)
in each of the two peers before gnunet\-tbench can be used.
.PP
The two peers must know of each other and be connected (use
gnunet\-stats to test for connections). Typically, gnunet\-tbench
reports the time it took to sent all specified messages and the
percentage of messages lost.
.PP

.TP
\fB\-c \fIFILENAME\fR, \fB\-\-config=\fIFILENAME\fR
load config file (defaults: ~/.gnunet/gnunet.conf)

.TP
\fB\-g \-\-gnuplot
create output in two colums suitable for gnuplot.
When using this option, concatenate the output of multiple
runs with various options into a file 'tbench' and run
the following gnuplot script to visualize the time/loss
ratio:
.TP
set xlabel "time"
set ylabel "percent transmitted"
plot "tbench" title 'Transport benchmarking' with points

.TP
\fB\-h\fR, \fB\-\-help\fR
print help page

.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=\fILOGLEVEL\fR
Change the loglevel.  Possible values for LOGLEVEL are NOTHING, FATAL,
ERROR, WARNING, INFO, STATUS and DEBUG.  Note that options in the 
configuration file take precedence over this option (the argument 
will be ignored in that case).

.TP
\fB\-i\fI ITER \fB\-\-iterations=\fIITER\fR
perform ITER iterations of the benchmark

.TP
\fB\-n\fI MESSAGES\fR, \fB\-\-msg=\fIMESSAGES\fR
how many messages should be used in each iteration (used to
compute average, min, max, etc.)

.TP
\fb\-r \fIRECEIVER\fR, \fB\-\-rec=\fIRECEIVER\fR
use this option to specify the identity of the
RECEIVER peer that is used for the benchmark. This option is required.

.TP
\fB\-s\fI SIZE \fB\-\-size=\fISIZE\fR
test using the specified message size

.TP
\fB\-S\fI SPACE \fB\-\-space=\fSPACE\fR
use SPACE milli\-seconds of delays between trains of messages

.TP
\fB\-t\fI TIMEOUT \fB\-\-timeout=\fTIMEOUT\fR
wait TIMEOUT milli\-seconds for replies to arrive before aborting

.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number

.TP
\fB\-X\fI COUNT \fB\-\-xspace=\fCOUNT\fR
use trains of COUNT messages

.SH NOTES
gnunet\-tbench can run for a long time, depending on
how high you have set the numbers. Run first with small numbers
to get an initial estimate on the runtime.
.SH FILES
.TP
~/.gnunet/gnunet.conf
Users GNUnet configuration file
.SH "REPORTING BUGS"
Report bugs by using mantis <https://gnunet.org/bugs/> or by sending electronic mail to <gnunet\-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunet.conf\fP(5), \fBgnunetd\fP(1)


ChangeLog
Mon Nov 16 09:25:47 CET 2009
	Allow passing of command-line arguments using GNUNET_ARGS
	environment variable (see Mantis #1467).

Mon Apr 27 22:18:19 CEST 2009
	Monitor low-traffic connections over a longer period
	of time before terminating them.

Fri Feb 27 22:05:50 MST 2009
	Releasing 0.8.0c.

Sun Feb  1 19:51:40 MST 2009
	Fixed Mantis #1429 (struct padding could cause sizeof
	to be different on AMD64, using gcc-ism "packed" to
	avoid this problem).

Sun Feb  1 00:37:16 MST 2009
	Fixed problem in DHT client protocol that could cause
	PUT requests from clients to be (frequently) lost.

Sun Nov  2 16:06:47 MST 2008
	Fixed problem with MySQL disconnects in one module
	impacting MySQL correctness in another module (caused
	crashes).

Sat Aug 16 00:18:34 CEST 2008
	Improved keyword extraction
	Fixed expiration time of keywords
	Releasing 0.8.0b.

Sun Jul 20 15:27:11 MDT 2008
	Releasing 0.8.0a.

Thu Jul 10 00:59:31 MDT 2008
	Fixed bug where datastore resizing would not 
	result in bloomfilter update during gnunet-update.

Sat Jun 28 23:43:00 MDT 2008
	Fixed bug preventing namespace advertisements from
	being processed.

Sat Jun 14 00:54:39 MDT 2008
	Releasing 0.8.0.

Fri Jun 13 23:51:00 MDT 2008
	GNUnet now compiles on OpenBSD.

Wed Jun 11 22:17:09 MDT 2008
	Made state of gnunet-auto-share persistent. 
	Added support for multiple directories.  Improved
	FSUI/ECRS APIs to reduce the number of threads needed
	for probing.  Increased number of concurrent probes
	allowed.

Sat Jun  7 17:54:49 MDT 2008
	Simplified namespaces a lot.  Shorter timeouts for
	trying to connect to gnunetd to prevent interactive
	apps from blocking too long when gnunetd is not
	running.

Sun Jun  1 21:11:54 MDT 2008
	Releasing 0.8.0pre1.

Sun Jun  1 13:41:27 MDT 2008
	Split of libgnunetpseudonym from libgnunetnamespace
	(with major API extension and changes).

Tue May  6 04:10:02 MDT 2008
	Added simple sanity check for [MODULES] values.
	Added "-s" and "-q" options to gnunet-peer-info.

Thu Apr 24 00:05:17 MDT 2008
	Simplified FSUI stop/abort/pause/resume API calls
	(no need to pass FSUI context anymore).

Mon Apr 21 21:05:20 MDT 2008
	Adding buffered IO for FSUI's serialize and
	deserialize code.

Sun Apr 20 20:01:20 MDT 2008
	Fixed bug in HTTP transport causing lonely messages
	(responses to GET) not to be delivered.

Mon Mar 24 21:15:36 MDT 2008
	Fixed various problems with downloading locally
	indexed large files (downloading large files from
	other peers should not have been a problem).
	
Sun Mar  2 15:33:41 MST 2008
	Added option to allow user to disable IPv6 support.
	Releasing 0.8.0pre0.

Sat Mar  1 21:14:02 MST 2008
	Fixed some dstore performance issues.
	Made transports alternate between trying IPv4 and IPv6
	if both are available (since one of them maybe
	misconfigured).

Mon Feb 25 23:25:48 MST 2008
	Cleanup of the DNS code: removed GNUNET_IPvXAddress
	and GNUNET_PRIP; centralized all DNS lookup code in 
	util/network/dns.c; improved IPv6 support for 
        network-client code.

Mon Feb 25 00:01:27 MST 2008
	Added asynchronous search and download methods for 
	ECRS library.  FSUI now can do with only one thread
	per search or download (until now, we had two 
	threads per search / download).

Tue Feb 19 20:35:28 MST 2008
	Updated database schemata to support O(1) operations
	even if there are N files under the same keywords.
	This breaks database backwards compatibility (and
	no migration code is provided).

Mon Feb 18 19:47:37 MST 2008
	Added new options -u and -s to gnunet-insert.

Sat Feb 16 21:37:33 MST 2008
	Implemented gnunet-auto-share for automatic
	sharing of directories.

Thu Feb 14 01:02:32 MST 2008
	Made HTTP transport work over IPv6 as well.

Wed Feb 13 22:59:07 MST 2008
	Modified mySQL and SQLite datastores to return diverse
	result sets during (partial) hash-based get iterations
	(should result in more diverse keyword result sets).

Tue Feb 12 23:54:34 MST 2008
	First shot at integrated hostlist HTTP server based
	on libmicrohttpd.

Sat Feb  9 22:06:23 MST 2008
	Minor changes to the ECRS/FSUI APIs for searches
	(no more timeouts and/or max-results).  Clients who
	need these features could easily implement them 
	directly (and they are not really useful to begin
	with).  
	Also, the IPv4 and IPv6 protocols are being integrated
	so that we only listen to one port (running both IPv4
	and IPv6 over the same port).  This break transport
	compatibility, but given that we're breaking protocol
	compatibility for file-sharing already, this may 
	actually be a bit of a good thing (since it will now
	be possible to tell if peers running the new GAP code
	are connected to other peers running the new GAP code).
	
Sun Feb  3 13:17:09 MST 2008
	Dramatic changes to the GAP implementation (breaking
	protocol compatibility).  Essentially, we can save
	a few bytes in each reply.  More importantly, the new
	code allows the searching client to specify a set of
	replies that are not desired (hopefully helping to
	dramatically increase the diversity of search replies
	obtained over time).  Note that the actual encoding
	and databases are not affected (just P2P protocol).
	The update is not complete yet, but should compile.

Tue Jan  8 20:07:20 MST 2008
	Added option for testing ("make check") to use weak(er)
	PRNG for key generation (thanks to Werner Koch for 
	pointing out how to do it).

Thu Dec 22 20:10:37 MST 2007
	Releasing GNUnet 0.7.3.

Sun Dec  9 14:34:32 MST 2007
	Implemented MySQL version of the dstore-module.  This
	means that sqLite is now again truly optional.

Sat Dec  8 15:15:53 MST 2007
	Integrated F2F topology into main topology module 
	(options are used to choose between F2F-only (old F2F), 
	 minimum number of friend connections (NEW) or entirely
	 arbitrary connection set (default)).

Thu Dec  6 20:51:17 MST 2007
	The SMTP transport is working again.

Thu Nov 22 20:49:16 MST 2007
	United all libgnunetutil_*.so libraries into one big
	libgnunetutil.so library (eliminates issues for binary

src/applications/tbench/gnunet-tbench.c
/*
     This file is part of GNUnet.
     (C) 2001, 2002, 2004, 2005, 2006 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file applications/tbench/gnunet-tbench.c
 * @brief Transport mechanism benchmarking tool
 * @author Paul Ruth, Christian Grothoff
 */

#include "platform.h"
#include "gnunet_directories.h"
#include "gnunet_protocols.h"
#include "gnunet_util.h"
#include "tbench.h"

#define DEFAULT_MESSAGE_SIZE  10
#define DEFAULT_TIMEOUT  	(2 * GNUNET_CRON_SECONDS)
#define DEFAULT_SPACING  	0

#define OF_HUMAN_READABLE 0
#define OF_GNUPLOT_INPUT 1

static unsigned long long messageSize = DEFAULT_MESSAGE_SIZE;

static unsigned long long messageCnt = 1;

static char *messageReceiver;

static unsigned long long messageIterations = 1;

static unsigned long long messageTrainSize = 1;

static GNUNET_CronTime messageTimeOut = DEFAULT_TIMEOUT;

static GNUNET_CronTime messageSpacing = DEFAULT_SPACING;

static int outputFormat = OF_HUMAN_READABLE;

static char *cfgFilename = GNUNET_DEFAULT_CLIENT_CONFIG_FILE;

/**
 * All gnunet-tbench command line options
 */
static struct GNUNET_CommandLineOption gnunettbenchOptions[] = {
  GNUNET_COMMAND_LINE_OPTION_CFG_FILE (&cfgFilename),   /* -c */
  GNUNET_COMMAND_LINE_OPTION_HELP (gettext_noop ("Start GNUnet transport benchmarking tool.")), /* -h */
  {'g', "gnuplot", NULL,
   gettext_noop ("output in gnuplot format"), 0,
   &GNUNET_getopt_configure_set_one, &outputFormat},
  GNUNET_COMMAND_LINE_OPTION_HOSTNAME,  /* -H */
  {'i', "iterations", "ITER",
   gettext_noop ("number of iterations"), 1,
   &GNUNET_getopt_configure_set_ulong, &messageIterations},
  GNUNET_COMMAND_LINE_OPTION_LOGGING,   /* -L */
  {'n', "msg", "MESSAGES",
   gettext_noop ("number of messages to use per iteration"), 1,
   &GNUNET_getopt_configure_set_ulong, &messageCnt},
  {'r', "rec", "RECEIVER",
   gettext_noop ("receiver host identifier (ENC file name)"), 1,
   &GNUNET_getopt_configure_set_string, &messageReceiver},
  {'s', "size", "SIZE",
   gettext_noop ("message size"), 1,
   &GNUNET_getopt_configure_set_ulong, &messageSize},
  {'S', "space", "SPACE",
   gettext_noop ("sleep for SPACE ms after each a message block"), 1,
   &GNUNET_getopt_configure_set_ulong, &messageSpacing},
  {'t', "timeout", "TIMEOUT",
   gettext_noop ("time to wait for the completion of an iteration (in ms)"),
   1,
   &GNUNET_getopt_configure_set_ulong, &messageTimeOut},
  GNUNET_COMMAND_LINE_OPTION_VERSION (PACKAGE_VERSION), /* -v */
  {'X', "xspace", "COUNT",
   gettext_noop ("number of messages in a message block"), 1,
   &GNUNET_getopt_configure_set_ulong, &messageTrainSize},
  GNUNET_COMMAND_LINE_OPTION_END,
};


/**
 * Tool to benchmark the performance of the P2P transports.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return return value from gnunetsearch: 0: ok, -1: error
 */
int
main (int argc, char *const *argv)
{
  struct GNUNET_ClientServerConnection *sock;
  CS_tbench_request_MESSAGE msg;
  CS_tbench_reply_MESSAGE *buffer;
  float messagesPercentLoss;
  struct GNUNET_GE_Context *ectx;
  struct GNUNET_GC_Configuration *cfg;
  int res;

  res = GNUNET_init (argc,
                     argv,
                     "gnunet-tbench",
                     &cfgFilename, gnunettbenchOptions, &ectx, &cfg);
  if (res == -1)
    {
      GNUNET_fini (ectx, cfg);
      return -1;
    }
  sock = GNUNET_client_connection_create (ectx, cfg);
  if (sock == NULL)
    {
      fprintf (stderr, _("Error establishing connection with gnunetd.\n"));
      GNUNET_fini (ectx, cfg);
      return 1;
    }

  msg.header.size = htons (sizeof (CS_tbench_request_MESSAGE));
  msg.header.type = htons (GNUNET_CS_PROTO_TBENCH_REQUEST);
  msg.msgSize = htonl (messageSize);
  msg.msgCnt = htonl (messageCnt);
  msg.iterations = htonl (messageIterations);
  msg.intPktSpace = GNUNET_htonll (messageSpacing);
  msg.trainSize = htonl (messageTrainSize);
  msg.timeOut = GNUNET_htonll (messageTimeOut);
  msg.priority = htonl (5);
  if (messageReceiver == NULL)
    {
      fprintf (stderr, _("You must specify a receiver!\n"));
      GNUNET_client_connection_destroy (sock);
      GNUNET_fini (ectx, cfg);
      return 1;
    }
  if (GNUNET_OK !=
      GNUNET_enc_to_hash (messageReceiver, &msg.receiverId.hashPubKey))
    {
      fprintf (stderr,
               _
               ("Invalid receiver peer ID specified (`%s' is not valid name).\n"),
               messageReceiver);
      GNUNET_free (messageReceiver);
      GNUNET_client_connection_destroy (sock);
      GNUNET_fini (ectx, cfg);
      return 1;
    }
  GNUNET_free (messageReceiver);

  if (GNUNET_SYSERR == GNUNET_client_connection_write (sock, &msg.header))
    {
      GNUNET_client_connection_destroy (sock);
      GNUNET_fini (ectx, cfg);
      return -1;
    }

  buffer = NULL;
  if (GNUNET_OK ==
      GNUNET_client_connection_read (sock,
                                     (GNUNET_MessageHeader **) & buffer))
    {
      GNUNET_GE_ASSERT (ectx,
                        ntohs (buffer->header.size) ==
                        sizeof (CS_tbench_reply_MESSAGE));
      if ((float) buffer->mean_loss < 0)
        {
          GNUNET_GE_BREAK (ectx, 0);
          messagesPercentLoss = 0.0;
        }
      else
        {
          messagesPercentLoss =
            (buffer->mean_loss / ((float) htons (msg.msgCnt)));
        }
      switch (outputFormat)
        {
        case OF_HUMAN_READABLE:
          printf (_("Time:\n"));
          PRINTF (_("\tmax      %llums\n"), GNUNET_ntohll (buffer->max_time));
          PRINTF (_("\tmin      %llums\n"), GNUNET_ntohll (buffer->min_time));
          printf (_("\tmean     %8.4fms\n"), buffer->mean_time);
          printf (_("\tvariance %8.4fms\n"), buffer->variance_time);

          printf (_("Loss:\n"));
          printf (_("\tmax      %u\n"), ntohl (buffer->max_loss));
          printf (_("\tmin      %u\n"), ntohl (buffer->min_loss));
          printf (_("\tmean     %8.4f\n"), buffer->mean_loss);
          printf (_("\tvariance %8.4f\n"), buffer->variance_loss);
          break;
        case OF_GNUPLOT_INPUT:
          printf ("%f %f\n", buffer->mean_time, 1.0 - messagesPercentLoss);
          break;
        default:
          printf (_("Output format not known, this should not happen.\n"));
        }
      GNUNET_free (buffer);
    }
  else
    printf (_
            ("\nDid not receive the message from gnunetd. Is gnunetd running?\n"));

  GNUNET_client_connection_destroy (sock);
  GNUNET_fini (ectx, cfg);
  return 0;
}

/* end of gnunet-tbench.c */
